jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1114 @@
+     */
+    @Override
+    public Set<E> edgesOf(V vertex)
+    {
+        synchronized (mutex) {
+            return c.edgesOf(vertex);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int inDegreeOf(V vertex)
+    {
+        synchronized (mutex) {
+            return super.inDegreeOf(vertex);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<E> incomingEdgesOf(V vertex)
+    {
+        synchronized (mutex) {
+            return c.incomingEdgesOf(vertex);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int outDegreeOf(V vertex)
+    {
+        synchronized (mutex) {
+            return super.outDegreeOf(vertex);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<E> outgoingEdgesOf(V vertex)
+    {
+        synchronized (mutex) {
+            return c.outgoingEdgesOf(vertex);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean removeAllEdges(Collection<? extends E> edges)
+    {
+        synchronized (mutex) {
+            return super.removeAllEdges(edges);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<E> removeAllEdges(V sourceVertex, V targetVertex)
+    {
+        synchronized (mutex) {
+            return super.removeAllEdges(sourceVertex, targetVertex);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean removeAllVertices(Collection<? extends V> vertices)
+    {
+        synchronized (mutex) {
+            return super.removeAllVertices(vertices);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean removeEdge(E e)
+    {
+        synchronized (mutex) {
+            if (c.removeEdge(e)) {
+                edgeSetModified();
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public E removeEdge(V sourceVertex, V targetVertex)
+    {
+        synchronized (mutex) {
+            E e = c.removeEdge(sourceVertex, targetVertex);
+            if (e != null)
+                edgeSetModified();
+            return e;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean removeVertex(V v)
+    {
+        synchronized (mutex) {
+            if (c.removeVertex(v)) {
+                edgeSetModified();
+                vertexSetModified();
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString()
+    {
+        synchronized (mutex) {
+            return super.toString();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<V> vertexSet()
+    {
+        synchronized (mutex) {
+            if (allVerticesSet != null)
```

Comment:
Because I noticed all the underlying implementation about this are lazy built. Deferred initialization may be more in line with the underlying design.

Reply:
Well, the downside of deferred initialization is that other methods have to keep checking whether this is null, which adds unnecessary overhead.  Since these are very lightweight objects, and vertexSet and edgeSet are likely to be called sooner rather than later by most algorithms, I think deferred initialization isn't worth it.